[Beck] Step2 - Actions change property

SangHwi-Back opened the pull request:
## 추가 작업 내용

* *Naming*에 대한 오해가 많이 있었던 코드입니다. Naming을 전체적으로 변경했습니다.
* UIView 에 확장되어 있던 view layer 관련 변수들이 Rectangle 뷰에서만 활성화 되도록 수정하였습니다.
* getter를 *private(set)* 접근제한자로 대신하였습니다.
* *RectRGBColor validation*을 위한 변수를 추가하였습니다.
* *입력 컨트롤러(ViewController)와 출력 컨트롤러(MainScreenViewController)를 나누고 그 사이를 Plane객체가 연결하도록 하였습니다.*   
*프로토콜을 이용한 Delegate패턴*을 통해 서로 통신할 수 있도록 하였습니다.
* *Rectangle에 convenience init을 사용*하여 지정 생성자를 재정의하는 것을 대체하였습니다.
* *UnitTest는 RectangleProperty만을 테스트*하도록 수정하였습니다.

## 추가 작업을 통해 느낀점

모델을 뷰와 분리하고, 모델을 관리하는 모델을 또 뷰 컨트롤로와 뷰에게서 분리하니 **기능수정이 원활합니다**. 처음에는 모델에 뷰와 관련된 부분을 완전히 분리하는 것이 나중의 나를 위한 작업이라는 것을 알게 되었습니다.

출력 컨트롤러라는 개념은 이번에 처음 사용해보았는데, ViewController와 Plane의 추상화가 더 잘 진행된다면 MainScreenViewController와 같이 다른 출력 컨트롤러 연결도 원활히 할 수 있을 것 같다는 생각이 들었습니다.

## 고민되는 사항

현재 Plane의 ```changeCurrentSelected(_:Rectangle?)``` 은 ```RectangleViewTapDelegate``` 프로토콜에 선언되어 있습니다. 그리고 내부에서 ```PlaneAdmitDelegate``` 프로토콜에 정의된 함수를 호출하고 있습니다.

델리게이트 패턴에 사용되는 프로토콜이 종속 관계가 된 것이 나중에 문제가 되지는 않을까 걱정입니다.

## Keyword

```MVC Pattern```

_아래 내용은 이전 PR 메시지입니다._
---

## 작업 내용

* 드로잉 앱의 요소들을 선택할 수 있고 색상과 알파 값을 변경할 수 있게 하였습니다.
* 현재 관리되고 있는 모델들(랜덤하게 만들어진 사각형)을 관리하는 ```Plane``` 객체를 구현하였습니다.
* 어노테이션(```@IBInspectable```, ```@IBInspectable```)을 이용하여 스토리보드 작업을 좀 더 수월하게 만들었습니다.
* 다크모드에서도 요소들이 잘 보이도록 separatorColor를 쓰는 등 색상 배치에 신경을 써 보았습니다.

## 아쉬운 점

코드가 testable하게 작성하지 못한 것 같습니다. 모델을 정확히 변경 가능한지는 테스트를 진행했습니다. 더 많은 케이스를 찾을 수 있었다면 좋았을 것 같다는 아쉬움이 남습니다.

```Plane``` 객체의 역할 자체도 많은 아쉬움이 남습니다. 저는 Plane 객체의 역할을 **오른쪽 패널이 알아야 할 정보와 액션을 수행하는 능력을 가진 객체** 로 정의하고 작업하였습니다. 하지만, 이 객체는 ```Model```같은데 모델이 너무 뷰와 결합이 강한 것 같다는 생각이 들어 많은 고민을 하게 되었습니다.

다음 과제가 MVC들의 각 요소들 결합을 조금 느슨하게 만드는 방법을 고민하는 것 같습니다. 다음 과제에서 Plane 객체를 추가 개선하도록 하겠습니다.

## 학습 키워드

```UITapGestureRecognizer``` ```MVC Pattern``` ```OOP``` ```XCTest```

![노트 2022  3  4](https://user-images.githubusercontent.com/65931336/156726550-4bbc4d3b-533c-4cff-a66e-ca71d60a0366.jpg)

SangHwi-Back committed:
CleanUp project files

불필요한 주석을 제거하고, gitignore 조건을 추가하였습니다.

SangHwi-Back committed:
Group Files of responsibility

모델과 로그 관련 파일들이 너무 많아서 관리가 힘들다고 판단하여 각각 폴더를 만들었습니다.

SangHwi-Back committed:
Definite Step2 tasks

스텝2에서 진행할 사항들을 미리 정의하고 작업을 시작합니다.

SangHwi-Back committed:
Update Storyboard environment

전체적인 뷰 구조를 생성하였습니다.
@IBDesignable, @IBInspectable 로 스토리보드에서 뷰의 Border를 정의할 수 있도록 하였습니다.

SangHwi-Back committed:
Update selecting view user interface

스크린 뷰 안에 생성된 뷰를 선택할 수 있는 유저 인터페이스를 구현합니다.
Plane 구조체에 필요한 인터페이스를 구현하였습니다.

SangHwi-Back committed:
Implement actions for randomView

생성되는 사각형들에 랜덤 색상과 투명도를 부여할 수 있는 인터페이스를 생성하였습니다.
현재 뷰 컨트롤러와 모델 간의 결합도가 높다고 판단되어 이 부분 수정하고
전체적인 리팩토링이 진행될 예정입니다.

SangHwi-Back committed:
Edit responsibility of Plane

Plane 객체에서 뷰 컨트롤러가 현재 선택한 뷰, 뷰의 프로퍼티를 가져올 수 있게 변경하였습니다.
모델 객체들을 전부 저장하고 있기 때문에, 현재 뷰와 상태를 가져오는 책임도 있다고 생각했습니다.
그 외에 Button, Slider의 이름이 동사라서 명사로 수정하였습니다.

SangHwi-Back committed:
Update Project in step2

README 파일을 개선하고 불필요한 파일은 제거하였습니다.

SangHwi-Back committed:
Test step2 and Refactor codes

모델 테스트를 위해 유닛 테스트 코드를 변경하고 관련 코드들도 수정하였습니다.
모델 관련 타입을 따로 파일 정의하여 떼어놓고, 모두 CustomStringConvertible 프로토콜을 구현했습니다.
일련의 작업들은 테스트를 용이하게 하기 위한 작업들이었습니다.

SangHwi-Back committed:
Update project before PR

AppCode를 이용하여 오탈자, 불필요한 표현 등을 선택적으로 제거하였습니다.

SangHwi-Back committed:
Correct alpha slider error

알파 값을 지정하는 슬라이더로 알파 값이 정확히 적용되지 않아서 수정

godrm commented on DrawingApp/DrawingApp/Models/FactoryViewRandomProperty.swift:
```diff
@@ -57,14 +56,14 @@ class FactoryViewRandomProperty: ViewPropertyCreator {
+        Double(getIntRandom(from: 1, to: 10))
     }
     
     struct ViewProperties {
```
모델을 선언할 때 그게 결과적으로 어떤 뷰에 반영이 될수는 있지만 이름에 View-가 다 붙어 있을 필요는 없습니다
그냥 그 도메인 의미와 역할을 충실하게 표현하는 단어가 더 좋습니다. View가 그걸 어디에 어떻게 가져다 쓸지는 모르는 거구요

godrm commented on DrawingApp/DrawingApp/Models/Plane.swift:
```diff
@@ -0,0 +1,63 @@
+    /// ViewController 내에 생성된 뷰 중 현재 선택된 뷰를 Plane 객체가 관리할 것입니다.
+    /// 뷰 선택이 가능하도록 하는 이유는 Plane 객체에서 저장된 프로퍼티의 값을 바꾸고
+    /// 그에 따라 바뀐 속성을 해당 뷰에 적용하기 위해서입니다.
+    var current: Rectangle?
```
현재 선택한 뷰를 외부에서 바꿔준다는 의미인가요? 그렇더라도 private(set)으로 할 수 있지 않을까요?

godrm commented on DrawingApp/DrawingApp/Models/Plane.swift:
```diff
@@ -0,0 +1,63 @@
+        properties.count
+    }
+    
+    func getRectangleProperty(at index: Int) -> ViewRandomProperty? {
+        guard properties.count-1 >= index else {
+            return nil
+        }
+        
+        return properties[index]
```
이 부분이 좀 어색한데, 왜 getRectangle이 아니라 getRectangleProperty인가요? 
Rectangle 이면 안되고 사각형의 속성들만 전달한다는 것을 강조한 건가요?

SangHwi-Back replied:
아... 이 부분은 설명이 부족했던 것 같습니다. 제가 Plane에서 사용한 ```Rectangle```들은 전부 **View**를 가리키고 있습니다.

그렇기 때문에 Rectangle이 되면 뷰 자체를 전달받는 함수라고 생각할 수 있어 네이밍을 ```getRectangleProperty(at:Int)```로 정의하게 되었습니다. 뒤에 **-Property**가 붙게 된 이유는 View가 아닌 ```Model```을 전달받는 함수이기 때문입니다.

PR 메시지에 말씀드린 대로 Plane 객체의 역할을 잘 이해하지 못해서 두루뭉술한 함수가 생겨난 것이 아닌가 싶습니다. 좀 더 코드를 면밀히 살펴보고 코드 개선하도록 하겠습니다.

godrm replied:
그러게요. 그런 정보를 PR 보낼 때 같이 설명하는 게 좋습니다. `Plane`에 뭐가 안 붙고 `Rectangle`에도 뭐가 안 붙어서 서로 같은 건 줄 알았는데 같은 역할이 아니고 하나는 모델이고 하나는 뷰였네요. 대비되는 모델과 뷰를 이름만 봐도 유추할 수 있도록 묶어주는 것은 중요한 포인트입니다.

godrm commented on DrawingApp/DrawingApp/Models/Plane.swift:
```diff
@@ -0,0 +1,63 @@
+        return properties[index]
+    }
+    
+    func hasAnyRectangle(in rect: RectPoint) -> Bool {
```
RectPoint 는 사각형 포인트처럼 보이는데 그러면 좌표가 하나가 아니라 4개인가요? 
hasAnyRectangle() 이 메소드는 특정 사각형 좌표 4개를 Plane 평면이 포함하고 있는지 판단해주는 함수가 맞나요?

SangHwi-Back replied:
RectPoint는 Coordinate System에서 설명하는 좌상단의 좌표 하나를 뜻하고 있습니다. 말씀하신 의미로도 해석될 수 있겠다고 생각이 들어 Naming을 수정해보도록 하겠습니다.

godrm commented on DrawingApp/DrawingApp/Models/Plane.swift:
```diff
@@ -0,0 +1,63 @@
+    
+    // MARK: - Edit properties (ViewContoller use properties)
+    func setProperty(at index: Int, alpha: Float) {
+        guard properties.count-1 >= index else { return }
+        properties[index].setAlpha(Double(alpha))
```
Plane 과 Rectangle 역할 분담을 잘 모르겠네요 
Rectangle 타입을 선언했다면 index가 아니라 Rectangle 타입으로 비교할 수는 없을까요?

SangHwi-Back replied:
Plane의 역할을 잘 이해하지 못한 것 같습니다. **Plane은 Model** 이라고 생각하고 작업을 진행하다 보니 

최대한 Plane에는 Rectangle에 관여하는 부분은 최소화 하고, Rectangle의 프로퍼티 역할을 하는 ViewRandomProperty를 다루도록 작업을 진행하였습니다.

확실히 **Rectangle을 관리하는 Plane** 으로 만들면 좀 더 명확한 책임을 가지는 Plane 객체가 만들어질 것 같습니다. 말씀해주신 방향으로 작업 진행해보도록 하겠습니다.

godrm replied:
Plane은 Model 맞습니다. 만약 같은 계층의 뷰를 만든다면 PlaneView나 CanvasView 나 뭐든 있어야겠죠
Plane에서 View를 처리하는 것이 있다면 안되는 거구요. 그걸 구분해보시면 될 것 같습니다

godrm commented on DrawingApp/DrawingApp/Models/RandomPropertySubTypes.swift:
```diff
@@ -0,0 +1,55 @@
+// MARK: - Rect Size, Point
+
+struct Rect {
+    struct Size {
+        var width: Double
+        var height: Double
+    }
+
+    struct Point {
+        var x: Double
+        var y: Double
+    }
```
아.. Rect.Point는 그냥 Point 좌표 하나인데 Rect.Point가 되면서 사각형 꼭지점 4개를 의미하는 줄 알았습니다

SangHwi-Back replied:
네이밍이 여기서 또..... 기가 막힌 네이밍을 생각해보겠습니다!

godrm commented on DrawingApp/DrawingApp/Models/RandomPropertySubTypes.swift:
```diff
@@ -0,0 +1,55 @@
+// MARK: - RectRGBColor
+
+struct RectRGBColor {
+    var r: Double
+    var g: Double
+    var b: Double
```
이렇게 RectRGBColor 값 타입으로 선언하면 RectRGBColor(r: -1234, g: 9999, b: 1234.1234) 같은 값도 생성이 되는건가요?
타입을 선언한다는 것은 생성할 때 안전하게 완전하게 생성이 되는지 정도는 보장을 해줘야 합니다
특히 범위가 있는 경우는 범위를 안전하게 생성할 수 있는지 init을 제공해주는 것도 고려해보시구요.

godrm commented on DrawingApp/DrawingApp/Models/ViewRandomProperty.swift:
```diff
@@ -0,0 +1,91 @@
+
+import Foundation
+
+class ViewRandomProperty: ViewPropertyCreator {
```
ViewRandomProperty 타입 이름을 보니 Random이 너무 강조된 것처럼 보이네요.
우리가 일부 속성을 랜덤하게 생성하기는 하지만, 그게 임의의 값인지 아닌지 타입에까지 명시할 필요가 있나요?
이 타입만 놓고 보면 랜덤으로 생성하는 게 제공되는 것도 아니니까요

godrm commented on DrawingApp/DrawingApp/Models/ViewRandomProperty.swift:
```diff
@@ -0,0 +1,91 @@
+    }
+    
+    @discardableResult
+    func setAlpha(_ alpha: Double) -> Bool {
+        guard alpha >= 0 else { return false }
+        self.alpha = alpha
+        return true
+    }
+    
+    func getAlpha() -> Double {
+        alpha
+    }
+    
+    @discardableResult
+    func setSize(_ size: RectSize) -> Bool {
+        guard size.width >= 0 || size.height >= 0 else { return false }
+        self.size = size
+        return true
+    }
+    
+    func getSize() -> RectSize {
+        size
+    }
+    
+    @discardableResult
+    func setPoint(_ point: RectPoint) -> Bool {
+        guard point.x >= 0 || point.y >= 0 else { return false }
+        self.point = point
+        return true
+    }
+    
+    func getPoint() -> RectPoint {
+        point
+    }
```
getter - setter를 명시적으로 제공하기 보다는 private(set) 정도를 더 우선적으로 고려해보고 
그래도 안되면 struct - let 정도로 값 타입으로 선언해도 됩니다. 그럴려고 값 타입을 사용하는 거니까요

godrm commented on DrawingApp/DrawingApp/TypeExtensions/UIView+Extension.swift:
```diff
@@ -0,0 +1,42 @@
+        set {
+            layer.cornerRadius = newValue
+        }
+    }
```
모든 뷰가 접근 가능하면 언젠가는 쓰겠지만, 그게 지금이 아니면 필요한 뷰에서만 제공해주는 것도 좋습니다
미래에 필요할 꺼 같아서 확장하는 것은 좋은 선택이 아닙니다. 
UIView를 상속받은 모든 뷰가 boder나 corner가 항상 잘 적용되는 것도 아닙니다. 스크롤뷰나 테이블뷰처럼 복잡한 뷰들은 오히려 제대로 동작하기 않기도 합니다.

godrm commented on DrawingApp/DrawingApp/ViewController.swift:
```diff
@@ -8,31 +8,80 @@
 import UIKit
 import os.log
 
+protocol RectangleViewTapDelegate {
+    func setSelected(_ view: Rectangle)
+}
+
+protocol RectangleViewValueChangeDelegate {
+    func setColor()
+    func setAlpha(as value: CGFloat)
+}
```
이 델리게이트 프로토콜들은 어떤 타입이 제공하는 델리게이트인가요?
그러면 선언도 제공하는 타입의 파일에 선언하는 게 좋을 것 같습니다
보통은 사용하는 쪽이 아니라 제공하는 쪽이 더 관련이 높으니까요

SangHwi-Back replied:
```RectangleViewValueChangeDelegate``` 는 현재 사용되지 않는 프로토콜입니다. 제외하도록 하겠습니다.

```RectangleViewTapDelegate``` 는 생성되는 View인 Rectangle의 터치 제스쳐가 실행될 경우 ViewController에 신호를 보내주기 위해 선언한 것입니다. **RectangleViewTapDelegate**를 Rectangle로 보내서 ViewController를 좀 더 보기좋게 바꾸도록 하겠습니다.

godrm replied:
다시 보니 `RectangleViewTapDelegate` 이름에서도 RectangleView라고 붙어있는데 실제 뷰는 Rectangle 이었으니 차이가 좀 있네요. 
View에만 뷰를 붙이거나 Model에는 모델을 붙이거나, 어느 한쪽에만 안 붙이거나 일관되게 이름을 붙여줘야 할 것 같습니다.

godrm commented on DrawingApp/DrawingApp/ViewController.swift:
```diff
@@ -8,31 +8,80 @@
+    }
+    
+    @IBAction func buttonForAddTouchUpInside(_ sender: UIButton) {
+        if let property = factory.makeRandomView(as: "subview #\(screenView.subviews.count)") {
```
이런 경우는 guard let으로 표현하고 실패하면 early exit 으로 빠져나가고 바인딩 하면 아래는 들여쓰지 않고 표현할 수 있습니다. 
근데 `"subview #\(screenView.subviews.count)"` 이 부분은 무슨 표현인지 모르겠네요. 
팩토리에 어떤 의도로 넘겨주는 거죠? 로그를 남기는 거였던가요? 기능과 로그는 좀 분리해도 좋을 것 같습니다

SangHwi-Back replied:
```"subview #\(screenView.subviews.count)"```는 해당 **뷰의 이름**입니다. 이전 프로그래밍 요구사항에서 뷰의 프로퍼티들을 단지 print할 때 예시에서 이름으로 보여지는 Rect0, Rect1, Rect2...이 있어서 넣은 것입니다. 기능과는 아무 상관이 없습니다...

현재는 이름을 제외해도 아무런 상관이 없고, 로그 관련 기능은 이전에 말씀해주신 대로 Cocoapods 라이브러리로 만들어 보기 위해 작업중입니다. 라이브러리가 어떻게 만들어질지도 아직 확실하지는 않으니 이름은 제외하도록 하겠습니다.

godrm replied:
아.. Rect0 같은 이름이었군요. 그건 괜찮은 것 같습니다. 생성할 때 이름이 필요할 테니까요

godrm commented on DrawingApp/DrawingApp/ViewController.swift:
```diff
@@ -8,31 +8,80 @@
-            let randomView = UIView(frame: CGRect.useViewModel(point: viewModel.point, size: viewModel.size))
-            randomView.backgroundColor = UIColor.useViewModel(rgbValue: viewModel.rgbValue, alpha: viewModel.alpha)
+            let rectangle = Rectangle(model: property)
+            rectangle.delegate = self
+            rectangle.index = plane.getRectangleCount()-1
             
-            LoggerUtil.faultLog(message: String(describing: viewModel))
+            screenView.addSubview(rectangle)
```
여기처럼 입력에 대한 흐름은 입력 -> 컨트롤러 -> plane까지 모델로 들어가기만 하면 됩니다
아래부분처럼 뷰를 생성하고 추가하는 것은 분리해보세요. 모델 -> 컨트롤러 -> 뷰 흐름을 따로 만들어야 합니다

godrm commented on DrawingApp/DrawingApp/ViewController.swift:
```diff
@@ -8,31 +8,80 @@
+            return
         }
+        
+        let color = property.resetRGBColor()
+        buttonSetRandomColor.backgroundColor = UIColor(
+            red: color.r/255,
+            green: color.g/255,
+            blue: color.b/255,
+            alpha: property.getAlpha()
+        )
+        
+        plane.current?.setBackgroundColor(using: property)
```
여기서도 새로운 컬러값을 plane의 현재 사각형에 전달하는 것으로 끝내고,
버튼의 색을 바꾸는 출력은 다른 흐름으로 구분해보세요

SangHwi-Back replied:
어제 불현듯 외국의 출처 미상의 밈처럼 자기 전에 저의 뇌가 생각해낸 것인데,

ViewController ---(컬러값)---> Plane ---(컬러값)---> ContainerView의 뷰컨트롤러
로 진행되는 흐름을 만드는 방식으로 진행해보겠습니다.

MVC에 대한 강의를 진행해주신 부분에서 Presentation Controller를 언급해주신 부분이 있는데,
스토리보드에서 ContainerView를 생성할 때 만들어지는 뷰 컨트롤러와 Plane객체가 통신할 수 있도록 만들어서 객체 간 책임을 나눠보도록 하겠습니다.

godrm commented on DrawingApp/DrawingApp/ViewController.swift:
```diff
@@ -8,31 +8,80 @@
+    
+    @IBAction func sliderAdmitAlphaValueChanged(_ sender: UISlider) {
+        
+        guard let index = plane.current?.index else { return }
+        
+        sender.value = round(sender.value)
+        plane.setProperty(at: index, alpha: sender.value)
+        buttonSetRandomColor.backgroundColor = buttonSetRandomColor.backgroundColor?.withAlphaComponent(CGFloat(sender.value/10))
+        plane.current?.setValue(alpha: sender.value)
```
여기도 입력 흐름으로 plane까지 전달하는 것과 뷰를 바꾸는 것을 분리해보세요
근데 이상한 게 plane에 값을 두 번 전달해야 하는건가요? 한 번만 전달하면 되지 않을까요?

godrm commented on DrawingApp/DrawingApp/ViewController.swift:
```diff
@@ -43,6 +92,44 @@ extension ViewController: MasterViewDelegate {
 
+extension ViewController: RectangleViewTapDelegate {
+    func setSelected(_ view: Rectangle) {
+        plane.current?.isSelected = false
+        
+        guard let property = plane.getRectangleProperty(at: view.index) else {
+            return
+        }
+        
+        buttonSetRandomColor.backgroundColor = view.backgroundColor
+        sliderSetAlpha.setValue(Float(round(property.getAlpha())), animated: true)
+        
+        view.isSelected = true
+        plane.current = view
```
setSelected()는 특정 뷰가 터치됐을 때 호출되는 것 맞죠? select가 동사인데 setSelected()는 선택을 받는 느낌이긴 하네요. 
plane 호출이 3번이나 있는데 간단하게 한 번만 호출할 수 있지 않을까 생각해보세요
터치 이벤트 입력이기 때문에 어떤 좌표를 눌렀는지 전달만 하면 되지 않을까요? Plane 내부에서 할 일이 상위로 올라온 듯한 느낌입니다
그리고 view라는 속성이 화면에 보이는 뷰인지, 보여줄 값을 가지는 모델인지 헷갈리네요. plane에 view를 저장하는건 아니니까요

SangHwi-Back replied:
ViewController.swift 71-76 line 에서처럼 Plane 의 일을 Plane이 처리하도록 수정하겠습니다. 네이밍은 계속 메모하면서 검토 진행중입니다...

godrm commented on DrawingApp/DrawingApp/Views/Rectangle.swift:
```diff
@@ -0,0 +1,75 @@
+import Foundation
+import UIKit
+
+class Rectangle: UIView {
```
Rectangle 이 UIView 타입이었군요
관련있는 뷰와 모델의 짝을 좀 더 명확하게 구분해줬으면 좋겠습니다
위에서 읽을때까지 모델인줄 알았네요. 그렇다면 Plane에서 Rectangle을 참조하는 것은 잘못된 구조입니다.
Plane에서는 UIKit과 독립적인 타입과 값들만 갖고 있어야 합니다

godrm commented on DrawingApp/DrawingApp/Views/Rectangle.swift:
```diff
@@ -0,0 +1,75 @@
+import Foundation
+import UIKit
+
+class Rectangle: UIView {
+    
+    var index: Int = 0
```
뷰의 index는 혹시 Plane에 포함된 순서 같은걸까요? 
index라는 것은 뷰끼리 순서라서 뷰를 포함하는 상위 모델이 관리하는 값이지, 뷰 자체가 가져야 하는 값이 아닌 것 같습니다

SangHwi-Back replied:
말씀해주신대로 Plane이 가질 수 있도록 소스 수정을 진행하도록 하겠습니다.

godrm commented on DrawingApp/DrawingApp/Views/Rectangle.swift:
```diff
@@ -0,0 +1,75 @@
+    var delegate: RectangleViewTapDelegate?
+    
+    required init?(coder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
```
수업에서 얘기한 것처럼 View 클래스의 init은 항상 init(frame:) 과 init(coder:) 둘 다 지원을 해줘야 합니다 
이렇게 fatalError()로 놔두시면 안 됩니다. 구현하라는 의미에요

godrm commented on DrawingApp/DrawingApp/Views/Rectangle.swift:
```diff
@@ -0,0 +1,75 @@
+        fatalError("init(coder:) has not been implemented")
+    }
+    
+    init(model: ViewRandomProperty) {
+        
+        let origin = model.getPoint()
+        let size = model.getSize()
+        
+        super.init(frame: CGRect(x: origin.x, y: origin.y, width: size.width, height: size.height))
+        
+        let tapGesture = UITapGestureRecognizer()
+        tapGesture.delegate = self
+        addGestureRecognizer(tapGesture)
+        
+        setBackgroundColor(using: model)
```
뷰는 지정 생성자를 새로 만드는 것을 매우 조심해야 합니다. 
그렇지 않으면 일반적인 형태가 되지 않아서 사용하기 어려운 타입이 되어 버립니다. 
기존의 지정 생성자를 활용할 수 있는 게 더 좋습니다.

godrm commented on DrawingApp/DrawingApp/Views/Rectangle.swift:
```diff
@@ -0,0 +1,75 @@
+        setBackgroundColor(using: model)
+    }
+    
+    func setModel(_ model: ViewRandomProperty) {
+        let origin = model.getPoint()
+        let size = model.getSize()
+        frame = CGRect(x: origin.x, y: origin.y, width: size.width, height: size.height)
+        setBackgroundColor(using: model)
+    }
+    
+    func setValue(alpha: Float) {
+        backgroundColor = backgroundColor?.withAlphaComponent((CGFloat(alpha / 10)))
+    }
+    
+    func setBackgroundColor(using model: ViewRandomProperty) {
+        
+        let color = model.getRGBColor()
+        
+        backgroundColor = UIColor(
+            red: color.r/255,
+            green: color.g/255,
+            blue: color.b/255,
+            alpha: model.getAlpha()/10
+        )
+    }
```
setBackgroundColor()과 setModel()는 중복된 느낌입니다. 필요한 값만 설정하도록 하는 게 어떨까요?

godrm commented on DrawingApp/DrawingApp/Views/Rectangle.swift:
```diff
@@ -0,0 +1,75 @@
+    
+    func gestureRecognizer(
+        _ gestureRecognizer: UIGestureRecognizer,
+        shouldReceive touch: UITouch
+    ) -> Bool {
+        delegate?.setSelected(self)
```
UIGestureRecognizerDelegate 에서 이 메소드는 어떤 역할을 하는건가요?
탭 조건이 맞는지 안 맞는지 판단하기 이전에 호출되는지 아닌지 확인해보셔야 할 것 같습니다
델리게이트 메소드 중에서도 적절한 메소드가 무엇인가 찾는 것도 중요합니다. 다음 자료를 한 번 보세요 
https://developer.apple.com/documentation/uikit/touches_presses_and_gestures/supporting_gesture_interaction_in_your_apps

godrm commented on DrawingApp/DrawingAppTests/DrawingAppAddViewTests.swift:
```diff
@@ -9,20 +9,42 @@ import XCTest
         
         let factory = FactoryViewRandomProperty()
-        factory.delegate = ViewController()
+        let vc = storyboard.instantiateViewController(withIdentifier: "MainViewController")
+        vc.loadView()
+        
+        if let vc = vc as? MasterViewDelegate {
+            factory.delegate = vc
+        } else {
+            XCTFail("MainViewController not implementing MasterViewDelegate")
+        }
```
보통은 뷰 컨트롤러를 생성하지 않고 테스트하는 게 좋습니다. 그래서 더 하위에 타입들을 실행해보지 않고 테스트할 수 있으니까요
여기처럼 ViewController를 생성하는 경우는 UI Test라고 별도로 화면 단위로 테스트를 합니다. 
단위 테스트에서는 UIKit과 관련없이 타입 하나만 테스트하는 게 좋습니다. 
수업에서도 한 번 얘기한 적이 있는 데 `@testable import DrawingApp`를 해서 모든 파일을 연결해서 테스트하기 보다는
테스트하려면 파일만 타깃에 추가해서 테스트해보세요. 지금 테스트는 단위 테스트는 아니고 앱 테스트 수준으로 더 어렵고 복잡한 과정이 필요합니다. 단위 테스트에 더 집중해보셨으면 합니다

godrm reviewed (changes requested):
살펴보다 보니 제가 잘 이해하지 못한 부분들이 있네요. 
타입 이름을 오해하기도 하고, 그러다보니 구조를 바로 파악하기 어려웠습니다
다음 피드백을 확인해보시고, 델리게이트 구조, Plane 모델과 뷰 관계를 개선해보셨으면 합니다.

SangHwi-Back committed:
Naming change for Rectangle related classes

ScreenView에 포함될 사각형 뷰의 네이밍이 전체적으로 이상하여 수정하였습니다.
불필요한 프로토콜도 제거하였습니다.

SangHwi-Back committed:
Refactor sourcecode

private(set)을 이용하여 getter를 삭제하였습니다.
일부 소스의 위치가 부적절하여 옮기게 되었습니다.
RectRGBColor의 값이 잘못 들어올 경우를 대비한 valication을 추가하였습니다.

SangHwi-Back committed:
Rectangle use convenience init

convenience init을 이용하여 기존의 Designated init을 사용하도록 수정하였습니다.

SangHwi-Back committed:
Create Present View Controller

사각형을 보여주고, 사각형 혹은 배경이 탭 될때 일어나는 액션을 관리하는 뷰 컨트롤러를 따로 생성하였습니다.
이를 통해 MVC Pattern에 좀 더 가까운 DrawingApp을 구축할 수 있었습니다.

SangHwi-Back committed:
Make documentation and final classes

좀 더 수월한 코드리딩을 위해 클래스 혹은 프로토콜 등에 문서화를 진행하였습니다.
또한, final 이라고 판단되는 클래스에는 final을 명시하였습니다.

SangHwi-Back committed:
Fix alpha value for Rectangle

사각형의 알파 값을 슬라이더로 변경한 후 다른 색상으로 바꾸게 되면,
슬라이더로 변경한 값이 유실되는 오류를 개선하였습니다.

SangHwi-Back committed:
Remove unnecessary procedures

탭 제스쳐에 selector들로 포함되었던 함수들이 필요하지 않아 제거하였습니다.

SangHwi-Back committed:
Make RectangleProperty testable

모델 클래스인 RectangleProperty 클래스만을 테스트에서 사용할 수 있도록 수정하였습니다.

SangHwi-Back committed:
Final touch before PR

PR을 다시 요청하기 전 코드리딩에 도움이 될만한 요소들을 추가하고,
전체적으로 코드를 재점검하였습니다.
색상을 재설정 하는 버튼의 색상이 앱의 동작과 일관성이 있지 않아서 기능을 수정하였습니다.

SangHwi-Back committed:
Separate Plane from View

Plane 객체에 뷰 관련 import와 소스가 들어갔는데 불필요하여 제거하였습니다.

SangHwi-Back committed:
Remove useless code

tapGesture를 생성하고 활성화 시키는 과정은 스토리보드에서 가능합니다.
관련 코드를 MainScreenViewController에서 삭제합니다.

godrm commented:
수정이 완료된 건지 댓글로 확인해주세요 @SangHwi-Back

SangHwi-Back commented:
아 죄송합니다. 기능이나 디자인 패턴 등에는 영향이 가진 않지만 추가로 하면 좋을 것 같아서 푸쉬해놓고 말씀을 안드렸습니다.

수정완료하였습니다. 다시 한번 확인 부탁드립니다! @godrm

godrm commented on DrawingApp/DrawingApp/Models/Plane.swift:
```diff
@@ -7,36 +7,46 @@
-    /// 그에 따라 바뀐 속성을 해당 뷰에 적용하기 위해서입니다.
+    private var properties = [RectangleProperty]()
+    /// Plane객체는 해당 변수를 이용해 선택된 뷰, 선택이 해제될 뷰를 관리합니다.
     var current: Rectangle?
```
Rectangle 이 View고 RectangleProperty가 Model 인거죠?
그렇다면 Plane은 View인 Rectangle을 참조하면 안됩니다. Plane이 더 하위 타입이고 UIKit에 의존하면 안되니까요
여기서는 모든 처리가 RectangleProperty를 기반으로 동작해야 합니다. 
그 부분을 고려해서 다음 단계에서는 분리해보세요.

godrm commented on DrawingApp/DrawingApp/Models/RandomPropertySubTypes.swift:
```diff
@@ -7,47 +7,89 @@
+    var allValues: [Double] { [r, g, b] }
+    static var maxValue: Double = 255
+    static var maxAlpha: Double = 10
+    private var _r: Double
+    var r: Double {
+        get {
+            _r
+        }
+        set {
+            if 0...RectRGBColor.maxValue ~= newValue {
+                _r = newValue
+            }
+        }
+    }
+    
+    private var _g: Double
+    var g: Double {
+        get {
+            _g
+        }
+        set {
+            if 0...RectRGBColor.maxValue ~= newValue {
+                _g = newValue
+            }
+        }
+    }
+    
+    private var _b: Double
+    var b: Double {
+        get {
+            _b
+        }
+        set {
+            if 0...RectRGBColor.maxValue ~= newValue {
+                _b = newValue
+            }
+        }
+    }
```
엄청 복잡한 struct RectRGBColor 타입이 되어 버렸네요
struct 는 class 와 다르게 var 속성을 바꾸면 메모리가 바뀌면서 새로운 값이 생성됩니다. 
그러다보면 값 타입으로 장점을 잃어버리고 클래스처럼 동작하게 되거든요
그래서 struct로 선언할 때는 struct let 형태로 불변 값 타입으로 만드는 게 가장 좋은 선택입니다. 
그럴려면 이렇게 computed property보다는 init에서 적당한 범위로 생성되고, 필요하면 또 다른 값을 만들어내도록 하는게 좋은거죠
자주 바뀐다면 차라리 클래스로 var를 만드는 게 낫습니다. 
이렇게 Computed property가 반복되는 것도 하위 타입이거나 property wrapper를 만드는 방법도 있구요.  

정리하면 복잡하게 만들거나, 심플하게 만들거나 선택해야 한다면 심플하게 값만 갖도록 만드는 게 더 좋을 수 있습니다
참고하세요

godrm commented:
수고하셨습니다. 
Plane은 모델이라서 직접 UIKit을 import 하지 않았지만, Rectangle을 쓰기 때문에 UIKit을 사용하는 셈입니다. 
그걸 분리해보는 게 중요한 도전이 될 것 같습니다.

SangHwi-Back commented:
고견 감사드립니다. 실수가 반복되는 느낌이 들어서 더 면밀히 검토하고 PR을 해야겠다는 생각이 드는군요...

다음 PR 때 뵙겠습니다.
